prover/src/circuits/summa_circuit.rs

```diff
@@ -84,28 +80,68 @@ impl<
 
         let username = meta.advice_column();
 
+        let concatenated_balance = meta.advice_column();
+        meta.enable_equality(concatenated_balance);
+
+        meta.create_gate("Concatenated balance sumcheck gate", |meta| {
+            let current_balance = meta.query_advice(concatenated_balance, Rotation::cur());
+            vec![current_balance.clone()]
+        });
+
+        let q_enable = meta.complex_selector();
+
         let balances = [(); N_CURRENCIES].map(|_| meta.advice_column());
         for column in &balances {
             meta.enable_equality(*column);
         }
 
-        meta.create_gate("Balance sumcheck gate", |meta| {
-            let mut nonzero_constraint = vec![];
-            for balance in balances {
-                let current_balance = meta.query_advice(balance, Rotation::cur());
-                nonzero_constraint.push(current_balance.clone());
+        meta.create_gate("Concatenated balance validation check gate", |meta| {
+            let s = meta.query_selector(q_enable);
+
+            let concatenated_balance = meta.query_advice(concatenated_balance, Rotation::cur());
+
+            // Right-most balance column is for the least significant balance in concatenated balance.
+            let mut balances_expr = meta.query_advice(balances[N_CURRENCIES - 1], Rotation::cur());
+
+            // Base shift value for 84 bits.
+            let base_shift = Fp::from(1 << 63).mul(&Fp::from(1 << 21));
+
+            let mut current_shift = Expression::Constant(base_shift);
+
+            for i in (0..N_CURRENCIES - 1).rev() {
+                let balance = meta.query_advice(balances[i], Rotation::cur());
+                let shifted_balance = balance * current_shift.clone();
+                balances_expr = balances_expr + shifted_balance;
+
+                if i != 0 {
+                    current_shift = current_shift * Expression::Constant(base_shift);
+                }
```

Comment:
This works because we know exactly that `N_CURRENCIES=3`, but will break if the prover changes the `N_CURRENCIES`. I think it's better to add an assertion about the parameter combination (i.e., would there be "overlap" between shifted values for the given N_USERS, N_CURRENCIES and max possible range of a balance) and also calculate the `base_shift` based on these parameters.

Reply:
While adding the assertions you suggested, I noticed that the verification of proof generated with `N_CURRENCIES` = 2 failed. I still don't have any clue why it doesn't work with two currencies. Therefore, I have simply modified it to allow only 1 or 3 currencies with the assertion.

I believe I can remove contract and fix the backend as I did in PR #297 after merging PR #292.